File: apps/frontend/rgbank/src/models/Expense.ts
import type { AccountBankInformation } from '@/models/AccountBankInformation'
import type { Category } from '@/models/Form'
import type { BookedItemData, ExpenseStatus } from '@/models/General'
import type { Thread } from '@/models/Thread'
import type { Committee, Student } from '@medieteknik/models'

export interface ExpenseData {
  files: File[]
  title: string
  description: string
  date: Date
  isDigital: boolean
  categories: Category[]
}

export interface ExpenseResponse {
  expense_id: string
  file_urls?: string[]
  title: string
  description: string
  date: string
  isDigital?: boolean
  categories?: Category[]
  status: ExpenseStatus
  created_at: string
  committee?: Committee
  amount?: number
  student?: Student
  booked_item?: BookedItemData
}

export interface ExpenseResponseDetailed {
  expense: ExpenseResponse
  student: Student
  bank_information: AccountBankInformation
  thread?: Thread
}

File: apps/backend/core-api/dto/apps/rgbank/expense.py
from typing import Any, Dict

from pydantic import BaseModel

from dto.apps.rgbank.thread import ThreadDTO
from dto.core.student import StudentDTO
from models.apps.rgbank.expense import PaymentStatus


class ExpenseDomainDTO(BaseModel):
    expense_part_id: str
    title: str
    parts: list[str]
    committee_id: str


class UpdateExpenseDomainDTO(BaseModel):
    title: str | None = None
    parts: list[str] | None = None


class ExpenseDTO(BaseModel):
    expense_id: str
    file_urls: list[str]
    title: str
    description: str
    date: str
    is_digital: bool
    categories: list[Dict[str, Any]]
    status: PaymentStatus
    created_at: str
    amount: float
    student: StudentDTO | None = None


class UpdateItemForm(BaseModel):
    status: str | None = None
    updatedCategories: list[Dict[str, Any]] | None = None
    comment: str | None = None


class CreateExpenseForm(BaseModel):
    files: list[bytes]
    date: str
    title: str
    description: str
    is_digital: str
    categories: list[Dict[str, Any]]


class InvoiceDTO(BaseModel):
    invoice_id: str
    already_paid: bool
    file_urls: list[str]
    title: str
    description: str
    is_original: bool
    is_booked: bool
    date_issued: str | None = None
    due_date: str | None = None
    categories: list[Dict[str, Any]]
    status: PaymentStatus
    created_at: str
    amount: float
    student: StudentDTO | None = None


class CreateInvoiceForm(BaseModel):
    files: list[bytes]
    already_paid: str
    title: str
    description: str
    is_original: str
    is_booked: str
    date_issued: str
    due_date: str
    categories: list[Dict[str, Any]]


class BaseItemResponseDTO(BaseModel):
    student: StudentDTO
    bank_information: str
    thread: ThreadDTO


class ExpenseResponseDTO(BaseItemResponseDTO):
    expense: ExpenseDTO


class InvoiceResponseDTO(BaseItemResponseDTO):
    invoice: InvoiceDTO
